src/api/l_headset.c

```diff
@@ -578,3 +587,93 @@ int luaopen_lovr_headset(lua_State* L) {
   headsetRenderData.ref = LUA_NOREF;
   return 1;
 }
+
+// TEMPORARY HANDS LIBRARY
+
+#include "headset/oculus_mobile_bridge.h"
+#include "headset/oculus_mobile.h"
+
+static void pushTableOfNumbers(lua_State* L, int idx, int count, ...) { // Takes floats
+  va_list args;
+  va_start(args, count);
+
+  lua_newtable(L);
+  for(int c = 0; c < count; c++) {
+    // Note we pass floats but they become doubles because of some garbage about how va_list works.
+    // This is probably okay becuase lua_number is probably a double anyway.
+    lua_pushnumber(L, va_arg(args, double));
+    lua_rawseti (L, -2, c+1);
+  }
+  lua_rawseti (L, -2, idx);
+
+  va_end(args);
+}
+
+static int l_lovrHeadsetHandsIsHand(lua_State* L) {
+  Device device = luax_optdevice(L, 1);
+  lovrAssert(device == DEVICE_HAND_LEFT || device == DEVICE_HAND_RIGHT, "Only works with hands");
+  LovrOculusMobileHands *hand = &lovrOculusMobileHands[device == DEVICE_HAND_RIGHT];
+
+  int points = 0;
+  lua_pushboolean(L, hand->live);
+
+  return 1;
+}
+
+static int l_lovrHeadsetHandsGetPoints(lua_State* L) {
```

Comment:
This function returns ALL joint poses. Each joint is a table containing {{x,y,z}, {a,b,c,d}} where abcd are the quaternion components (so, it's not angle axis like Lovr poses normally are). The order of the joint poses is hardcoded and is the order VrApi lists them in. In my personal code, I make list of this ordering with these two slightly awkward tables.

```
HandCubes.skeletonSkeleton = {
  [1]=2, -- Forearm
  [3]=1, [4]=3, [5]=4, [6]=5, [20]=6, -- Thumb
  [7]=1, [8]=7, [9]=8, [21]=9,        -- Index
  [10]=1, [11]=10, [12]=11, [22]=12,  -- Middle
  [13]=1, [14]=13, [15]=14, [23]=15,  -- Ring
  [16]=1, [17]=16, [18]=17, [19]=18, [24]=19 -- Pinky
}
HandCubes.skeletonTipStart = 20
HandCubes.expectedLength = 24
HandCubes.skeletonFingers = {
	{4, 5, 6, 20}, -- Thumb
	{7, 8, 9, 21},    -- Index
	{10, 11, 12, 22}, -- Middle
	{13, 14, 15, 23}, -- Ring
	{17, 18, 19, 24}, -- Pinky
}
```

Points are returned in "hand space". To get global space you need to compose the normal `lovr.headset.getPose` pose. In my personal code I get the list of global-space points with:

```
		local points = tablex.map(
			function(point)
				return Loc( vec3(unpack(point[1])), quat(unpack(point[2])) ):compose(handAt)
			end,
			lovr.headset.hand.getPoints(controllerName)
		)
```

There was some kind of problem I hit early on involving the left hand being rotated 180 degrees all the time but I don't remember where I resolved it.

Reply:
Parent-space joint poses might make more sense for the bridge interface.  Composing the parent space transforms into higher spaces is more straightforward than going backwards from hand/global transforms (e.g. matrix multiplications vs. matrix inversions).  Also `Model:pose` takes poses in parent space, so using parent space would mean that the poses could be directly copied to a model without conversions.

It might be less overall work too, since the [OVR_Skeleton code](https://github.com/mcclure/lovr-oculus-mobile/blob/lovr-oculus-mobile-hand/VrSamples/VrHands/Src/OVR_Skeleton.cpp#L62) used to generate this is already seemingly taking the parent space poses from vrapi and converting them to hand space for the bridge.  It seems wasteful to take parent-space poses, unconditionally convert them to hand space, then invert them back to parent space.  Instead they could remain in parent space and only get converted to hand/global space if that's what the Lua code wants to do.

Reply:
This makes a lot of sense to me.